### PR TITLE
Market-Peer merge with server

### PR DESCRIPTION
**Pull request by xiej2520**
# Description

Merges the market with the peer, and starts implementing the HTTP server.
This uses Sea Horses' market, as it appears to be the most featureful implementation we have.

This also forks from https://github.com/daminals/orcanet-rust/tree/jobs-endpoints

## List of changes

* Merges Sea Horses' market with the peer (currently tracking changes as they are merged by [reaovyd])
* Implements [file chunk hashing protocol](https://docs.google.com/document/d/1IfiCf2OMkDiqluHIxmNPOc2y4UB7WDJXR_fflhfca0A/edit#heading=h.hrwx6rigxcwc)
* Related to previous point, modified the producer to use `FileInfoHash` newtypes everywhere instead of `String`ly typed objects for consistency and type safety.
* Moved `jobs.rs` and `files.rs` to a transfer module, following [this suggestion](https://github.com/daminals/orcanet-rust/pull/6/commits/c29ef67e2daeeed04c854ab94f346c234f26a013).
* Organized routes by team and by section on the API doc.
* Started implementing routes, includes the ones from the `jobs-endpoints` branch, as well as add/start/pause/terminate job.

There is a local market provided (compile and use with `cargo build --bin peer-node-server -F test_local_market`)

## Todo

* Implement rest of routes
* Integrate fully working market when completed
* Coin integration

### CODE REVIEW

**Comment by xiej2520** [comment]
@daminals @andrewctam @reaovyd

Been pushing most of my changes to this branch (has the most api endpoints implemented).
We could try merging it, but I think the jobs implementation is currently a mess, any concerns about organization?

File: .github/workflows/peernode.yml [hunk 1]
```diff
@@ -38,11 +38,25 @@ jobs:
...
+#        sleep 5 &&
+#        cargo test seapig_test --verbose -- --test-threads=1 --ignored
+#    - name: Start server and run bubble guppies tests
+#      working-directory: peernode
+#      run: |
+#        cargo build --bin peer-node-server -F test_local_market &&
+#        cargo run --bin peer-node-server -F test_local_market &
+#        sleep 10 &&
+#        cargo test bubbleguppies_test --verbose -- --test-threads=1 --ignored
+
```
Comment (daminals) [inline .github/workflows/peernode.yml]: What does this do?

File: peernode/src/peer-node/routes/bubble_guppies/settings.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+#![allow(non_snake_case)]
```
Comment (andrewctam) [inline peernode/src/peer-node/routes/bubble_guppies/settings.rs]: These endpoints were removed in the latest spec

File: .github/workflows/peernode.yml (hunk 1 above)
Comment (xiej2520) [reply .github/workflows/peernode.yml]: I was attempting to run the tests on Github Actions by starting a server and then running a test but shelved it for now. The tests work locally if you start up a server and then run tests, not sure what's going on with github.

File: peernode/src/peer-node/routes/file_routes.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
+#![allow(non_snake_case)]
```
Comment (andrewctam) [inline peernode/src/peer-node/routes/file_routes.rs]: We can delete this file if the routes are implemented in sea pig's directory

File: .github/workflows/peernode.yml (hunk 1 above)
Comment (daminals) [reply .github/workflows/peernode.yml]: Ok, so just add a TODO comment for now here

**Comment by daminals** [comment]
> @daminals @andrewctam @reaovyd
> 
> Been pushing most of my changes to this branch (has the most api endpoints implemented). We could try merging it, but I think the jobs implementation is currently a mess, any concerns about organization?

I think this is fine, but the PR should be made in the main 23sp repo with the appropriate Closes x tags to improve visibility and also to break up the PRs (since making a pr with the entire main branch would be way too hard to review)

**Comment by andrewctam** [comment]
The server is really deep inside the peernode directory, maybe we can take the transfer and routes all the way out into a server mod at the root (we can change transfers name, it was named that originally due to conflicts with server.rs)

File: server/src/jobs.rs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+            JobStatus::Active(_) => write!(f, "active"),
+            JobStatus::Paused(_) => write!(f, "paused"),
+            JobStatus::Stop => write!(f, "stopping"),
+            JobStatus::Completed => write!(f, "completed"),
+            JobStatus::Failed => write!(f, "failed"),
+        }
+    }
+}
+
+pub async fn start(job: AsyncJob, token: String) {
```
Comment (andrewctam) [inline server/src/jobs.rs]: Should we move this down to impl Jobs? Once the job is finished, it needs to call self.finish_job to add it to the history.

Comment (xiej2520) [reply server/src/jobs.rs]: The task that's spawned needs to release the lock on job in order for other tasks to possibly grab the lock and pause the task (which is why it's called on `AsyncJob`). There's probably a better way to do this, or maybe `start`, `pause`, `terminate` could be moved to methods of `Jobs`.

File: server/src/routes/bubble_guppies/jobs.rs [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+        let jobs = state.jobs.lock().await;
+
+        match jobs.get_job(&job_id).await {
+            Some(job) => {
+                let token = config.get_token(job.lock().await.encoded_producer.clone());
+
+                jobs::start(job, state.jobs.clone(), token).await;
+            }
+            None => {
+                return (StatusCode::NOT_FOUND, format!("Job {job_id} not found")).into_response()
```
Comment (andrewctam) [inline server/src/routes/bubble_guppies/jobs.rs]: Is start_jobs intended to end prematurely if one of the jobs is not found? pause_jobs and terminate_jobs is counting failures and will look at all the jobs even if one fails.

Comment (xiej2520) [reply server/src/routes/bubble_guppies/jobs.rs]: Nope, that should be corrected.

**Review by andrewctam — APPROVED** [review]

**Review by daminals — APPROVED** [review]

**Pull request merged**